### PR TITLE
feat: Add support for Spark-compatible cast from integral to decimal

### LINKED ISSUE #2049: Implement Spark-compatible cast from integral types to decimal
### What is the problem the feature request solves?

When casting from intergral types to decimal, we currently delegate to DataFusion if `COMET_CAST_ALLOW_INCOMPATIBLE` config is enabled. The DataFusion implementation is not compatible with Spark.

We should implement a Spark-compatible cast implementation within Comet instead. Once this is implemented, we should also update the native `is_datafusion_spark_compatible` function to no longer return true for casts from integer to decimal.

There is an existing test in `CometCastSuite` which is currently ignored and should be enabled (and updated as required). 

### Describe the potential solution

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by coderfender**
## Which issue does this PR close?

Closes #2049 

## Rationale for this change

These changes are implemented to support numeric types to Decimal 128 types

## What changes are included in this PR?

Implement comet native cast operations to cast integral types to decimals. 

## How are these changes tested?

1. Re-enable long and Int -> Decimal128 cast tests
2. Add more tests to check overflow / ANSI edge cases

### CODE REVIEW

**Comment by coderfender** [comment]
@andygrove  , requesting CI kickoff

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala [hunk 1]
```diff
@@ -322,8 +322,7 @@ class CometCastSuite extends CometTestBase with AdaptiveSparkPlanHelper {
     castTest(generateInts(), DataTypes.DoubleType)
   }
 
-  ignore("cast IntegerType to DecimalType(10,2)") {
-    // Comet should have failed with [NUMERIC_VALUE_OUT_OF_RANGE] -1117686336 cannot be represented as Decimal(10, 2)
+  test("cast IntegerType to DecimalType(10,2)") {
```
Comment (parthchandra) [inline spark/src/test/scala/org/apache/comet/CometCastSuite.scala:325]: Not sure if this will test for conditions where we will get overflow/error in the native code

Comment (coderfender) [reply spark/src/test/scala/org/apache/comet/CometCastSuite.scala:325]: Great point . Let me go ahead and add some tests to check overflow conditions as well

Comment (coderfender) [reply spark/src/test/scala/org/apache/comet/CometCastSuite.scala:325]: @parthchandra  , I looked into [DataGenerator.scala](https://github.com/apache/datafusion-comet/blob/eea40ca1526fcbf7c6b811bba8d28289b00be09f/spark/src/test/scala/org/apache/comet/DataGenerator.scala#L99) and it seems like we are testing for random int values along with adding `Int.MIN` and `Int.MAX` values which are well beyond `Decimal(10,2)` 's precision .  I even added a special test case (not committed to the branch yet) but I almost feel it is redundant at the moment. Please advice if you think it is a good idea to add the below test to explicitly verify overflow values only (which passes with no issues):

```
  test("cast IntegerType to DecimalType(10,2) overflowable check") {
    val intToDecimal10OverflowValues = withNulls(Seq(Int.MinValue, -100000000, -100000001, 100000000, 100000001, Int.MaxValue)).toDF("a")
    castTest(intToDecimal10OverflowValues, DataTypes.createDecimalType(10, 2))
  }
```

Comment (parthchandra) [reply spark/src/test/scala/org/apache/comet/CometCastSuite.scala:325]: I think we can add this test, even if it is only to make it explicitly clear that we are testing this condition as well.

Comment (coderfender) [reply spark/src/test/scala/org/apache/comet/CometCastSuite.scala:325]: Sure @parthchandra  . Let me push a commit with that test added. I hope everything is else is OK with the change ?

Comment (andygrove) [reply spark/src/test/scala/org/apache/comet/CometCastSuite.scala:325]: I agree, let's add explicit tests

**Comment by andygrove** [comment]
This is looking good @coderfender. I will review more thoroughly later this week.

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala (hunk 1 above)
Comment (coderfender) [reply spark/src/test/scala/org/apache/comet/CometCastSuite.scala:325]: @parthchandra , @andygrove . I added tests below per the suggestion

**Comment by comphead** [comment]
Once its done, I'll check it with TPCH

**Comment by coderfender** [comment]
> Once its done, I'll check it with TPCH

Thank you

File: native/spark-expr/src/conversion_funcs/cast.rs [hunk 2]
```diff
@@ -1464,6 +1465,104 @@ where
...
+                        builder.append_value(scaled);
+                    }
+                }
+                _ => {
+                    return Err(SparkError::NumericValueOutOfRange {
+                        value: v.to_string(),
+                        precision,
+                        scale,
+                    })
+                }
```
Comment (andygrove) [inline native/spark-expr/src/conversion_funcs/cast.rs]: eval_mode needs to be respected here as well:

```suggestion
                _ => match eval_mode {
                    EvalMode::Ansi => {
                        return Err(SparkError::NumericValueOutOfRange {
                            value: v.to_string(),
                            precision,
                            scale,
                        });
                    }
                    EvalMode::Try | EvalMode::Legacy => builder.append_null(),
                },
```

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala [hunk 3]
```diff
@@ -322,11 +322,16 @@ class CometCastSuite extends CometTestBase with AdaptiveSparkPlanHelper {
...
     castTest(generateInts(), DataTypes.createDecimalType(10, 2))
   }
 
+  test("cast IntegerType to DecimalType(10,2) overflow check") {
+    val intToDecimal10OverflowValues =
+      Seq(Int.MinValue, -100000000, -100000001, 100000000, 100000001, Int.MaxValue).toDF("a")
+    castTest(intToDecimal10OverflowValues, DataTypes.createDecimalType(10, 2))
+  }
```
Comment (andygrove) [inline spark/src/test/scala/org/apache/comet/CometCastSuite.scala:333]: This test does not test all of the native implementation code. For example, it did not test the case where the `checked_mul` returned `None`. I suggest testing with other precisions and scales, including `DecimalType.MAX_PRECISION` and `DecimalType.MAX_SCALE`.

Comment (coderfender) [reply spark/src/test/scala/org/apache/comet/CometCastSuite.scala:333]: Added tests to verify multiple decimal scales and precisions

File: native/spark-expr/src/conversion_funcs/cast.rs (hunk 2 above)
Comment (coderfender) [reply native/spark-expr/src/conversion_funcs/cast.rs]: Done. Thank you for the suggestion @andygrove

**Comment by coderfender** [comment]
@andygrove the one test which failed is perhaps due to a transient error. Can you please rerun that test only whenever you get a chance ?

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala [hunk 4]
```diff
@@ -1232,7 +1244,6 @@ class CometCastSuite extends CometTestBase with AdaptiveSparkPlanHelper {
             case (None, None) =>
             // neither system threw an exception
             case (None, Some(e)) =>
-              // Spark succeeded but Comet failed
```
Comment (andygrove) [inline spark/src/test/scala/org/apache/comet/CometCastSuite.scala:1235]: is it intentional that this comment is removed?

**Comment by coderfender** [comment]
Thank you   for the approval @andygrove

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala (hunk 4 above)
Comment (coderfender) [reply spark/src/test/scala/org/apache/comet/CometCastSuite.scala:1235]: Not really . Let me fix this minor issue

**Comment by andygrove** [comment]
> Once its done, I'll check it with TPCH

@comphead this is now merged

**Comment by coderfender** [comment]
@comphead , I would like to volunteer in helping with the TPCH benchmarks too
